[TypeName] Nested types should respect MaxNode count

adamsitnik opened the pull request:
Fixes a bug discovered by @jkotas in https://github.com/dotnet/runtime/pull/103713#discussion_r1714635958

fixes #106991

adamsitnik committed:
add a failing unit test

adamsitnik committed:
the fix

adamsitnik committed:
handle 100% of the cases: nested generic types, generics of nested generics

adamsitnik committed:
minor polishing after reading the code again

adamsitnik commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeNameParserHelpers.cs:
```diff
@@ -388,5 +394,18 @@ internal static void ThrowInvalidOperation_HasToBeArrayClass()
 #endif
         }
+
+        internal static bool IsMaxDepthExceeded(TypeNameParseOptions options, int depth)
```
The `TypeNameParserHelpers.cs` file is included as a link in the test project, so we can have plenty of small unit tests for each of the helper methods: https://github.com/dotnet/runtime/blob/main/src/libraries/System.Reflection.Metadata/tests/Metadata/TypeNameParserHelpersTests.cs

Because of that, it can not use internal methods of `TypeNameParseOptions`. That is why I've moved `IsMaxDepthExceeded` from `TypeNameParseOptions` to here.

adamsitnik commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeNameParserHelpers.cs:
```diff
@@ -388,5 +394,18 @@ internal static void ThrowInvalidOperation_HasToBeArrayClass()
+#if SYSTEM_PRIVATE_CORELIB
+            => false; // CoreLib does not enforce any limits
+#else
+            => depth > options.MaxNodes;
```
important: so far the check was greater of equal (`>=`). After wrapping my head around I got the conclusion that we should allow for the max value too (it feels more natural).

jkotas commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeNameParser.cs:
```diff
@@ -147,14 +147,31 @@ private TypeNameParser(ReadOnlySpan<char> name, bool throwOnError, TypeNameParse
+                // If that definition represents a nested type, this needs to be taken into account.
+                if (nestedNameLengths is not null)
+                {
+                    recursiveDepth += nestedNameLengths.Count;
```
Should this be done earlier as soon as we discover that the nesting is too deep?

Also, I am having hard time understanding why the `TryDive` at https://github.com/dotnet/runtime/pull/106334/files#diff-905542846dc45d4936395b08bdac3b71d1f5e03529a188298a93f20d080c19c3R252 is not sufficient to handle this case.

adamsitnik replied:
> Should this be done earlier as soon as we discover that the nesting is too deep?

Ideally yes, but at that moment of time we don't know whether the name that is being parsed represents a generic type. 

> Also, I am having hard time understanding why the TryDive at is not sufficient to handle this case.

This `TryDive` increases the node count once, but if it turns out that it's a generic type, we need to do that again (because the generic type definition is also going to be  a nested type and will require the same amount of type name instances.

I've added following comment to the code, please let me know if this is enough:

```cs
// If that generic type is a nested type, this needs to be taken into account. Example:
// "Namespace.Declaring+NestedGeneric`1[GenericArg]" requires 5 TypeName instances:
// - constructed generic type: Namespace.Declaring+NestedGeneric`2[GenericArg] (+1, handled by ParseNextTypeName)
//   - declaring type: Namespace.Declaring (+1, handled by TryGetTypeNameInfo)
//   - generic type definition: Namespace.Declaring+NestedGeneric`1 (+1, handled by the TryDive above)
//      - declaring type: Namespace.Declaring (+1, handled by the if below)
//   - generic arguments:
//     - simple: GenericArg (+1, handled by ParseNextTypeName)
```

adamsitnik commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeName.cs:
```diff
@@ -284,22 +284,25 @@ public int GetNodeCount()
-            }
-            else if (IsConstructedGenericType)
+            TypeName? declaring = _declaringType;
+            while (declaring is not null)
```
I added more tests, deleted the old implementation and started from scratch as I did not know where the bug was.

I am going to revert this particular change:

```diff
- TypeName? declaring = _declaringType;
- while (declaring is not null)
+ if (IsNested)
{
-    result++;
-    declaring = declaring._declaringType;
+    result += DeclaringType.GetNodeCount();
}
```

What turned out to be a bug was missing `GetGenericTypeDefinition().GetNodeCount();`

adamsitnik committed:
address code review feedback:

- revert the change that seemed like optimization to make the code more readable
- add a comment explaining why we don't need to check of integer overflows
- add a comment explaining the details of generic nested types node count

adamsitnik committed:
Merge remote-tracking branch 'upstream/main' into nestedTypeNameNodeCount

# Conflicts:
#	src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeNameParserHelpers.cs

jkotas commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeName.cs:
```diff
@@ -284,22 +284,25 @@ public int GetNodeCount()
+
+                foreach (TypeName genericArgument in GetGenericArguments())
+                {
+                    result += genericArgument.GetNodeCount();
```
`TypeName.Parse` is not the only way to create new type names after the last PR. Try something like this:
```
using System.Reflection.Metadata;
using System.Collections.Immutable;

var simpleTypeName = TypeName.Parse("A");
var tn = simpleTypeName;

for (int i = 0; i < 20; i++)
{
    tn = simpleTypeName.MakeGenericTypeName(ImmutableArray.Create(tn, tn, tn));
    Console.WriteLine(tn.GetNodeCount());
}
```
It is going to hit the overflow in this code.

jkotas commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeNameParser.cs:
```diff
@@ -147,14 +147,38 @@ private TypeNameParser(ReadOnlySpan<char> name, bool throwOnError, TypeNameParse
+                    return null;
+                }
+                // If that generic type is a nested type, this needs to be taken into account. Example:
+                // "Namespace.Declaring+NestedGeneric`1[GenericArg]" requires 5 TypeName instances:
```
`Namespace.Declaring+NestedGeneric`1[GenericArg]` should only require 4 TypeName instances.

I think this comment is not correct. Also, the implementation of GetNodeCount seems to have a bug - it is double-counting nested TypeName instances.

adamsitnik replied:
> `Namespace.Declaring+NestedGeneric[GenericArg]` should only require 4 TypeName instances.

Which node from the list below should not be included in your opinion?

```log
- constructed generic type: Namespace.Declaring+NestedGeneric`2[GenericArg] (+1, handled by ParseNextTypeName)
  - declaring type: Namespace.Declaring (+1, handled by TryGetTypeNameInfo)
  - generic type definition: Namespace.Declaring+NestedGeneric`1 (+1, handled by the TryDive above)
     - declaring type: Namespace.Declaring (+1, handled by the if below)
  - generic arguments:
    - simple: GenericArg (+1, handled by ParseNextTypeName)
```

We define it as the total number of `TypeName` instances required to represent it:

https://github.com/dotnet/runtime/blob/89839049b486bf48239dd6faa722defd3de320cb/src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeName.cs#L275-L276

And we have 5 instances.

![image](https://github.com/user-attachments/assets/32093c45-d5e4-4acf-aff7-0dfeb1904d21)

The declaring type of the generic type definition and of the constructed generic type is the same object:

![image](https://github.com/user-attachments/assets/4e3103c2-d9a5-40db-929d-244ca672cc95)

Do you suggest that because it's the same object, it should not be included?

> Also, the implementation of GetNodeCount seems to have a bug - it is double-counting nested TypeName instances.

It does that on purpose: to include the nodes from generic type definition.

jkotas replied:
> Do you suggest that because it's the same object, it should not be included?

Yes. If it is the same object, it is the same instance and it should be only counted once. 

We have multiple ways to get to that instance through the object model. I do not think that it is the reason to count it multiple times.

adamsitnik replied:
> Yes. If it is the same object, it is the same instance and it should be only counted once

You are right. I've changed the implementation and the tests, PTAL.

jkotas commented:
I have opened #106991 to make sure that this does not get missed for .NET 9.

adamsitnik commented:
> I have opened #106991 to make sure that this does not get missed for .NET 9.

Thank you for doing that and apologies for the delay in responding.

adamsitnik committed:
Merge remote-tracking branch 'upstream/main' into nestedTypeNameNodeCount

adamsitnik committed:
address code review feedback:

- handle overflows
- don't include the declaring type of generic type definition, as it's the same instance uses by the constructed generic type

and add sth extra: assert for ensuring that GetNodeCount returns the same result as calculated by the parser during parsing

jkotas commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeName.cs:
```diff
@@ -301,24 +301,29 @@ public string Name
             {
-                result += GetElementType().GetNodeCount();
-            }
+                checked { result++; } // the generic type definition
```
```suggestion
                result = checked(result + GetGenericTypeDefinition().GetNodeCount());
```
I think the logic would be easier to understand if it is written like this, and the `if (IsNested)` check is moved to `else if` after the `IsConstructedGenericType` check.

jkotas reviewed (commented):
LGTM otherwise

adamsitnik committed:
address code review feedback:

- reorder the conditions
- add test for int32 overflow

adamsitnik commented:
/azp list

adamsitnik commented:
/azp run runtime-libraries-coreclr outerloop-linux

jkotas commented on src/libraries/System.Reflection.Metadata/tests/Metadata/TypeNameTests.cs:
```diff
@@ -493,6 +581,26 @@ public void GetNodeCountReturnsExpectedValue(Type type, int expected)
+            TypeName genericType = TypeName.Parse("Generic".AsSpan());
+            TypeName genericArg = TypeName.Parse("Arg".AsSpan());
+            // Don't allocate Array.MaxLength array as it may make this test flaky (frequent OOMs).
+            ImmutableArray<TypeName>.Builder genericArgs = ImmutableArray.CreateBuilder<TypeName>(initialCapacity: int.MaxValue / 64);
```
This does not need to allocate very large array. Instead, it can repeat MakeGenericTypeName multiple times. It will make the test a lot less resource intensive, faster and compatible with 32-bit platforms.

adamsitnik replied:
> This does not need to allocate very large array. Instead, it can repeat MakeGenericTypeName multiple times.

I know and in my opinion this is exactly what it's doing right now. But I assume that `int.MaxValue / 64` is still very large array for you. What would not be? `int.MaxValue / 1024`?

jkotas replied:
It can be very small, like 10.

adamsitnik replied:
`Math.Sqrt(int.MaxValue)` does what we need, I've pushed the change

jkotas commented on src/libraries/System.Reflection.Metadata/tests/Metadata/TypeNameTests.cs:
```diff
@@ -493,6 +581,26 @@ public void GetNodeCountReturnsExpectedValue(Type type, int expected)
         }
 
+        [OuterLoop("It takes a lot of time")]
+        [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.Is64BitProcess))] // 32 is likely to OOM
```
```suggestion
```

adamsitnik committed:
address code review feedback: allocate smaller arrays so it can run on 32 bit

adamsitnik committed:
Merge remote-tracking branch 'upstream/main' into nestedTypeNameNodeCount

adamsitnik commented:
@MihuBot fuzz TypeName

jkotas reviewed (commented):
LGTM. Thank you!

adamsitnik commented:
/ba-g unrelated

jkotas commented:
@adamsitnik Are you going to port this to release/9.0?

adamsitnik commented:
> @adamsitnik Are you going to port this to release/9.0?

Yes, I am going to send multiple backport PRs (with the fuzzers, the bug fixes they discovered and this PR)

adamsitnik commented:
> @adamsitnik Are you going to port this to release/9.0?

Follow up: https://github.com/dotnet/runtime/pull/107533
